lib/util/usedPropTypes.js

```diff
@@ -0,0 +1,520 @@
+              node: properties[k]
+            });
+          }
+        }
+        break;
+      default:
+        break;
+    }
+
+    components.set(component ? component.node : node, {
+      usedPropTypes: usedPropTypes,
+      ignorePropsValidation: ignorePropsValidation
+    });
+  }
+
+  /**
+   * @param {ASTNode} node We expect either an ArrowFunctionExpression,
+   *   FunctionDeclaration, or FunctionExpression
+   */
+  function markDestructuredFunctionArgumentsAsUsed(node) {
+    const destructuring = node.params && node.params[0] && node.params[0].type === 'ObjectPattern';
+    if (destructuring && components.get(node)) {
+      markPropTypesAsUsed(node);
+    }
+  }
+
+  function handleSetStateUpdater(node) {
+    if (!node.params || node.params.length < 2 || !inSetStateUpdater()) {
+      return;
+    }
+    markPropTypesAsUsed(node);
+  }
+
+  /**
+   * Handle both stateless functions and setState updater functions.
+   * @param {ASTNode} node We expect either an ArrowFunctionExpression,
+   *   FunctionDeclaration, or FunctionExpression
+   */
+  function handleFunctionLikeExpressions(node) {
+    handleSetStateUpdater(node);
+    markDestructuredFunctionArgumentsAsUsed(node);
+  }
+
+  function handleCustomValidators(component) {
+    const propTypes = component.declaredPropTypes;
+    if (!propTypes) {
+      return;
+    }
+
+    Object.keys(propTypes).forEach(key => {
+      const node = propTypes[key].node;
+
+      if (astUtil.isFunctionLikeExpression(node)) {
+        markPropTypesAsUsed(node);
+      }
+    });
+  }
+
+  return {
+    VariableDeclarator: function(node) {
+      const destructuring = node.init && node.id && node.id.type === 'ObjectPattern';
+      // let {props: {firstname}} = this
+      const thisDestructuring = destructuring && node.init.type === 'ThisExpression';
+      // let {firstname} = props
+      const statelessDestructuring = destructuring && isPropAttributeName(node) && (
+        utils.getParentStatelessComponent() ||
+        isInLifeCycleMethod(node)
+      );
+
+      if (!thisDestructuring && !statelessDestructuring) {
+        return;
+      }
+      markPropTypesAsUsed(node);
+    },
+
+    FunctionDeclaration: handleFunctionLikeExpressions,
+
+    ArrowFunctionExpression: handleFunctionLikeExpressions,
+
+    FunctionExpression: handleFunctionLikeExpressions,
+
+    MemberExpression: function(node) {
+      if (isPropTypesUsage(node)) {
+        markPropTypesAsUsed(node);
+      }
+    },
+
+    ObjectPattern: function(node) {
+      // If the object pattern is a destructured props object in a lifecycle
+      // method -- mark it for used props.
+      if (isNodeALifeCycleMethod(node.parent.parent)) {
+        node.properties.forEach((property, i) => {
+          if (i === 0) {
+            markPropTypesAsUsed(node.parent);
+          }
+        });
+      }
+    },
+
+    'Program:exit': function() {
+      const list = components.list();
+
+      for (const component in list) {
```

Comment:
`for..in` should be avoided; can this use normal array iteration methods?

Reply:
Sure. `Object.keys(list).forEach` perhaps?
There's about a dozen other files already doing the same, handle them all?

Reply:
yes, that would be great!

Reply:
already done in https://github.com/yannickcr/eslint-plugin-react/pull/1946/commits/b35ea0c72317d4123237af32759f60858343b682 😄